src/bootstrap/config.rs

```diff
@@ -782,14 +771,75 @@ impl Config {
             set(&mut config.missing_tools, t.missing_tools);
         }
 
+        // This is duplicated in bootstrap.py.
+        config.out = build.build_dir.unwrap_or_else(|| PathBuf::from("build"));
+        t!(fs::create_dir_all(&config.out));
+        config.out = config.out.canonicalize().expect("Build directory canonicalized");
+        config.out = config
+            .out
+            .components()
+            .flat_map(|component| {
+                if component.as_os_str() == "$ROOT" {
+                    config.src.components().collect::<Vec<_>>()
+                } else {
+                    vec![component]
+                }
+            })
+            .collect();
+
+        if config.dry_run {
+            let dir = config.out.join("tmp-dry-run");
+            t!(fs::create_dir_all(&dir));
+            config.out = dir;
+        }
+
+        let build_rustc = build.rustc;
+        config.initial_rustc = env::var_os("RUSTC")
+            .map(PathBuf::from)
+            .or_else(|| {
+                // Try a rustc binary by the CARGO. Cargo doesn't provide an
+                // absolute path to us in $RUSTC for build scripts.
+                let rustc = config.initial_cargo.with_file_name(exe("rustc", config.build));
+                if rustc.exists() {
+                    if let Some(file_cfg) = build_rustc.clone().map(PathBuf::from) {
+                        assert_eq!(rustc, file_cfg.canonicalize().expect("rustc exists"));
+                    }
+                    Some(rustc)
+                } else {
+                    build_rustc.clone().map(PathBuf::from)
+                }
+            })
+            .or_else(|| build_rustc.clone().map(PathBuf::from))
+            .unwrap_or_else(|| panic!("Unable to discover a rustc binary"));
+
+        // Cargo does not provide a RUSTFMT environment variable, so we
+        // synthesize it manually. Note that we also later check the config.toml
+        // and set this to that path if necessary.
+        let rustfmt = config.initial_rustc.with_file_name(exe("rustfmt", config.build));
+        config.initial_rustfmt = if rustfmt.exists() { Some(rustfmt) } else { None };
+
         // Now that we've reached the end of our configuration, infer the
         // default values for all options that we haven't otherwise stored yet.
 
-        set(&mut config.initial_rustc, build.rustc.map(PathBuf::from));
-        set(&mut config.initial_cargo, build.cargo.map(PathBuf::from));
+        // Make sure that our variables, pulled from Cargo, are equivalent to
+        // configuration.
+        if let Some(cargo) = build.cargo.map(PathBuf::from) {
+            assert_eq!(config.initial_cargo, cargo.canonicalize().expect("cargo exists"));
```

Comment:
Could `initial_cargo` be set to the `CARGO` env var used to compile the build script?

Reply:
We do that here -- https://github.com/rust-lang/rust/pull/76544/files#diff-1731657fe6a8025cbe7a498b30e93890R478 -- this is just checking that *if* you've done that and also configured initial-cargo in the toml file they match. Maybe we shouldn't bother though and just treat the command line as an override?

Reply:
Ah yeah in that case I'd just treat this as an override.